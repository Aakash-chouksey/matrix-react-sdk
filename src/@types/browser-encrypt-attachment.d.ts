/*
Copyright 2021 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

declare module "browser-encrypt-attachment" {
    interface IEncryptedAttachment {
        data: ArrayBuffer;
        info: IEncryptedAttachmentInfo;
    }

    export interface IEncryptedAttachmentInfo {
        key: string;
        iv: string;
        hashes: {
            sha256: string;
        };
    }

    function encryptAttachment(plaintextBuffer: ArrayBuffer): Promise<IEncryptedAttachment>;
    function decryptAttachment(ciphertextBuffer: ArrayBuffer, info: IEncryptedAttachmentInfo): Promise<ArrayBuffer>;

    export { encryptAttachment, decryptAttachment };
}
