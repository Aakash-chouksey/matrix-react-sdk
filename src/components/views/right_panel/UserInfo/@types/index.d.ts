/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

export interface PowerLevelsContent {
    events?: Record<string, number>;
    // eslint-disable-next-line camelcase
    users_default?: number;
    // eslint-disable-next-line camelcase
    events_default?: number;
    // eslint-disable-next-line camelcase
    state_default?: number;
    ban?: number;
    kick?: number;
    redact?: number;
}

export interface RoomPermissions {
    modifyLevelMax: number;
    canEdit: boolean;
    canInvite: boolean;
}

export interface Device {
    deviceId: string;
    ambiguous?: boolean;
    getDisplayName(): string;
}
