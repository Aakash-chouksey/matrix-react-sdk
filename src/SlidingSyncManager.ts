/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/


/*
 * Sliding Sync Architecture - MSC https://github.com/matrix-org/matrix-spec-proposals/pull/3575
 * 
 * This is a holistic summary of the changes made to Element-Web / React SDK / JS SDK to enable sliding sync.
 * This summary will hopefully signpost where developers need to look if they want to make changes to this code.
 * 
 * At the lowest level, the JS SDK contains an HTTP API wrapper function in client.ts. This is used by
 * a SlidingSync class in JS SDK, which contains code to handle list operations (INSERT/DELETE/SYNC/etc)
 * and contains the main request API bodies, but has no code to control updating JS SDK structures: it just
 * exposes an EventEmitter to listen for updates. When MatrixClient.startClient is called, callers need to
 * provide a SlidingSync instance as this contains the main request API params (timeline limit, required state,
 * how many lists, etc).
 * 
 * The SlidingSyncSdk INTERNAL class in JS SDK attaches listeners to SlidingSync to update JS SDK Room objects,
 * and it conveniently exposes an identical public API to SyncApi (to allow it to be a drop-in replacement).
 * 
 * At the highest level, SlidingSyncManager contains mechanisms to tell UI lists which rooms to show,
 * and contains the core request API params used in Element-Web. It does this by listening for events
 * emitted by the SlidingSync class and by modifying the request API params on the SlidingSync class.
 * 
 *    (entry point)                     (updates JS SDK)
 *  SlidingSyncManager                   SlidingSyncSdk
 *       |                                     |
 *       +------------------.------------------+
 *         listens          |          listens
 *                     SlidingSync
 *                     (sync loop,
 *                      list ops)
 */

import { MatrixClient } from 'matrix-js-sdk';
import { MSC3575List, SlidingSync, SlidingSyncEvent, SlidingSyncState } from 'matrix-js-sdk/src/sliding-sync';
import { logger } from "matrix-js-sdk/src/logger";

// how long to long poll for
const SLIDING_SYNC_TIMEOUT_MS = 20 * 1000;

// the things to fetch when a user clicks on a room
const DEFAULT_ROOM_SUBSCRIPTION_INFO = {
    timeline_limit: 50,
    required_state: [
        ["*", "*"], // all events
    ]
};

/**
 * This class manages the entirety of sliding sync at a high UI/UX level. It controls the placement
 * of placeholders in lists, controls updating sliding window ranges, and controls which events
 * are pulled down when. The intention behind this manager is be the single place to look for sliding
 * sync options and code.
 */
export class SlidingSyncManager {
    slidingSync: SlidingSync;
    client: MatrixClient;

    constructor(client: MatrixClient, proxyUrl: string){
        this.slidingSync = new SlidingSync(proxyUrl, [], DEFAULT_ROOM_SUBSCRIPTION_INFO, client, SLIDING_SYNC_TIMEOUT_MS);
        this.client = client;
    }

    /**
     * Ensure that this list is registered.
     * @param listIndex The list index to register
     * @param updateArgs The fields to update on the list.
     * @returns The complete list request params
     */
    async ensureListRegistered(listIndex: number, updateArgs: { filters?: object, sort?: string[], ranges?: number[][] }): Promise<MSC3575List> {
        console.log("ensureListRegistered", listIndex, updateArgs);
        let list = this.slidingSync.getList(listIndex);
        if (!list) {
            list = {
                ranges: [ [0, 20] ],
                sort: [
                    "by_highlight_count", "by_notification_count", "by_recency",
                ],
                timeline_limit: 1, // most recent message display: though this seems to only be needed for favourites?
                required_state: [
                    ["m.room.join_rules", ""], // the public icon on the room list
                    ["m.room.avatar", ""], // any room avatar
                    ["m.room.tombstone", ""], // lets JS SDK hide rooms which are dead
                    ["m.room.encryption", ""], // lets rooms be configured for E2EE correctly
                    ["m.room.create", ""], // for isSpaceRoom checks
                    ["m.room.member", this.client.getUserId()], // lets the client calculate that we are in fact in the room
                ]
            };
            list = Object.assign(list, updateArgs);
        } else {
            const updatedList = Object.assign({}, list, updateArgs);
            if (JSON.stringify(list) === JSON.stringify(updatedList)) {
                console.log("list matches, not sending, update => ", updateArgs);
                return list;
            }
            list = updatedList;
        }
        this.slidingSync.setList(listIndex, list);

        return new Promise((resolve, reject) => {
            const resolveOnSubscribed = (state, resp, err) => {
                if (state === SlidingSyncState.Complete) { // we processed a /sync response
                    this.slidingSync.off(SlidingSyncEvent.Lifecycle, resolveOnSubscribed);
                    resolve(list);
                }
            };
            // wait until the next sync before returning as RoomView may need to know the current state
            this.slidingSync.on(SlidingSyncEvent.Lifecycle, resolveOnSubscribed);
        });
    }

    setRoomVisible(roomId: string, visible: boolean): Promise<string> {
        let subscriptions = this.slidingSync.getRoomSubscriptions();
        if (visible) {
            subscriptions.add(roomId);
        } else {
            subscriptions.delete(roomId);
        }
        logger.log("SlidingSync setRoomVisible:",roomId, visible);
        this.slidingSync.modifyRoomSubscriptions(subscriptions);

        return new Promise((resolve, reject) => {
            const resolveOnSubscribed = (state, resp, err) => {
                if (state === SlidingSyncState.Complete) { // we processed a /sync response
                    this.slidingSync.off(SlidingSyncEvent.Lifecycle, resolveOnSubscribed);
                    resolve(roomId);
                }
            };
            // wait until the next sync before returning as RoomView may need to know the current state
            this.slidingSync.on(SlidingSyncEvent.Lifecycle, resolveOnSubscribed);
        });
    }

};

let manager = null;
export function getSlidingSyncManager(): SlidingSyncManager {
    return manager;
}
export function create(client: MatrixClient, proxyUrl: string): SlidingSync {
    manager = new SlidingSyncManager(client, proxyUrl);
    return manager.slidingSync;
}