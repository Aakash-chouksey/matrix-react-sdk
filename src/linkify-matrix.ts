/*
Copyright 2015, 2016 OpenMarket Ltd
Copyright 2019 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import * as linkifyjs from 'linkifyjs';
import { registerPlugin } from 'linkifyjs';
import { baseUrl } from "./utils/permalinks/SpecPermalinkConstructor";
import {
    parsePermalink,
    tryTransformEntityToPermalink,
    tryTransformPermalinkToLocalHref,
} from "./utils/permalinks/Permalinks";
import { RoomMember } from 'matrix-js-sdk';
import dis from './dispatcher/dispatcher';
import { Action } from './dispatcher/actions';
import { ViewUserPayload } from './dispatcher/payloads/ViewUserPayload';

enum Type {
    URL = "url",
    UserId = "userid",
    RoomAlias = "roomalias",
    GroupId = "groupid"
}

// Linkify stuff doesn't type scanner/parser/utils properly :/
function matrixOpaqueIdLinkifyParser({
    scanner,
    parser,
    utils,
    token,
    name,
}: {
    scanner: any;
    parser: any;
    utils: any;
    token: '#' | '+';
    name: Type;
}) {
    const {
        DOMAIN,
        DOT,
        // A generic catchall text token
        TEXT,
        NUM,
        TLD,
        COLON,
        SYM,
        UNDERSCORE,
        // because 'localhost' is tokenised to the localhost token,
        // usernames @localhost:foo.com are otherwise not matched!
        LOCALHOST,
    } = scanner.tokens;

    const S_START = parser.start;
    const Localpart = utils.createTokenClass(name, { isLink: true });

    const localpartTokens = [
        DOMAIN,
        NUM,
        TLD,

        // because 'localhost' is tokenised to the localhost token,
        // usernames @localhost:foo.com are otherwise not matched!
        LOCALHOST,
        SYM,
        UNDERSCORE,
        TEXT,
    ];

    const HASH_STATE = S_START.tt(token);
    const LOCALPART_STATE = HASH_STATE.tt(DOMAIN, Localpart);

    for (const token of localpartTokens) {
        HASH_STATE.tt(token, LOCALPART_STATE);
        LOCALPART_STATE.tt(token, LOCALPART_STATE);
    }

    const TLD_STATE = LOCALPART_STATE.tt(TLD);
    TLD_STATE.tt(DOMAIN, LOCALPART_STATE);
    TLD_STATE.tt(TLD, LOCALPART_STATE);

    const DOT_STATE = LOCALPART_STATE.tt(DOT, TLD_STATE);

    const COLON_STATE = LOCALPART_STATE.tt(COLON);

    COLON_STATE.tt(LOCALHOST, LOCALPART_STATE);
    COLON_STATE.tt(TLD, LOCALPART_STATE);
    COLON_STATE.tt(DOMAIN, LOCALPART_STATE);

    LOCALPART_STATE.tt(COLON, COLON_STATE);
    LOCALPART_STATE.tt(DOT, DOT_STATE); // accept repeated TLDs (e.g .org.uk)
    const PORT_STATE = LOCALPART_STATE.tt(COLON);
    PORT_STATE.tt(NUM, LOCALPART_STATE);
}

function matrixUserIdLinkifyPlugin({ scanner, parser, utils }) {
    const {
        AT,
        DOMAIN,
        DOT,
        UNDERSCORE,
        EQUALS,
        MINUS,
        SLASH,
        NUM,
        TLD,
        COLON,
        // because 'localhost' is tokenised to the localhost token,
        // usernames @localhost:foo.com are otherwise not matched!
        LOCALHOST,
    } = scanner.tokens;

    const S_START = parser.start;
    const UserId = utils.createTokenClass(Type.UserId, { isLink: true });

    // Slightly more defined tokens list than for OpaqueID
    const localpartTokens = [
        DOMAIN,
        NUM,
        TLD,
        NUM,
        DOT,
        UNDERSCORE,
        EQUALS,
        MINUS,
        SLASH,

        // because 'localhost' is tokenised to the localhost token,
        // usernames @localhost:foo.com are otherwise not matched!
        LOCALHOST,
    ];

    const HASH_STATE = S_START.tt(AT);
    const LOCALPART_STATE = HASH_STATE.tt(DOMAIN, UserId);

    for (const token of localpartTokens) {
        HASH_STATE.tt(token, LOCALPART_STATE);
        LOCALPART_STATE.tt(token, LOCALPART_STATE);
    }

    const TLD_STATE = LOCALPART_STATE.tt(TLD);
    TLD_STATE.tt(DOMAIN, LOCALPART_STATE);
    TLD_STATE.tt(TLD, LOCALPART_STATE);

    const DOT_STATE = LOCALPART_STATE.tt(DOT, TLD_STATE);

    const COLON_STATE = LOCALPART_STATE.tt(COLON);

    COLON_STATE.tt(LOCALHOST, LOCALPART_STATE);
    COLON_STATE.tt(TLD, LOCALPART_STATE);
    COLON_STATE.tt(DOMAIN, LOCALPART_STATE);

    LOCALPART_STATE.tt(COLON, COLON_STATE);
    LOCALPART_STATE.tt(DOT, DOT_STATE); // accept repeated TLDs (e.g .org.uk)
    const PORT_STATE = LOCALPART_STATE.tt(COLON);
    PORT_STATE.tt(NUM, LOCALPART_STATE);
}

export const matrixLinkify: Record<any, any> = {
    // stubs, overwritten in MatrixChat's componentDidMount
    onUserClick: function(e: MouseEvent, userId: string) {
        const member = new RoomMember(null, userId);
        if (!member) { return; }
        dis.dispatch<ViewUserPayload>({
            action: Action.ViewUser,
            member: member,
        });
    },
    onAliasClick: function(e: MouseEvent, roomAlias: string) {
        event.preventDefault();
        dis.dispatch({ action: 'view_room', room_alias: roomAlias });
    },
    onGroupClick: function(e: MouseEvent, groupId: string) {
        event.preventDefault();
        dis.dispatch({ action: 'view_group', group_id: groupId });
    },
};

const escapeRegExp = function(string): string {
    return string.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
};

// Recognise URLs from both our local and official Element deployments.
// Anyone else really should be using matrix.to.
export const ELEMENT_URL_PATTERN =
    "^(?:https?://)?(?:" +
        escapeRegExp(window.location.host + window.location.pathname) + "|" +
        "(?:www\\.)?(?:riot|vector)\\.im/(?:app|beta|staging|develop)/|" +
        "(?:app|beta|staging|develop)\\.element\\.io/" +
    ")(#.*)";

export const MATRIXTO_URL_PATTERN = "^(?:https?://)?(?:www\\.)?matrix\\.to/#/(([#@!+]).*)";
export const MATRIXTO_MD_LINK_PATTERN =
    '\\[([^\\]]*)\\]\\((?:https?://)?(?:www\\.)?matrix\\.to/#/([#@!+][^\\)]*)\\)';
export const MATRIXTO_BASE_URL= baseUrl;

export const options = {
    events: function(href: string, type: Type | string): Partial<GlobalEventHandlers> {
        switch (type) {
            case Type.URL: {
                // intercept local permalinks to users and show them like userids (in userinfo of current room)
                try {
                    const permalink = parsePermalink(href);
                    if (permalink && permalink.userId) {
                        return {
                            // @ts-ignore see https://linkify.js.org/docs/options.html
                            click: function(e) {
                                matrixLinkify.onUserClick(e, permalink.userId);
                            },
                        };
                    }
                } catch (e) {
                    // OK fine, it's not actually a permalink
                }
                break;
            }
            case Type.UserId:
                return {
                    // @ts-ignore see https://linkify.js.org/docs/options.html
                    click: function(e) {
                        matrixLinkify.onUserClick(e, href);
                    },
                };
            case Type.RoomAlias:
                return {
                    // @ts-ignore see https://linkify.js.org/docs/options.html
                    click: function(e) {
                        matrixLinkify.onAliasClick(e, href);
                    },
                };
            case Type.GroupId:
                return {
                    // @ts-ignore see https://linkify.js.org/docs/options.html
                    click: function(e) {
                        matrixLinkify.onGroupClick(e, href);
                    },
                };
        }
    },

    formatHref: function(href: string, type: Type | string): string {
        switch (type) {
            case Type.RoomAlias:
            case Type.UserId:
            case Type.GroupId:
            default: {
                return tryTransformEntityToPermalink(href);
            }
        }
    },

    linkAttributes: {
        rel: 'noreferrer noopener',
    },

    target: function(href: string, type: Type | string): string {
        if (type === Type.URL) {
            try {
                const transformed = tryTransformPermalinkToLocalHref(href);
                if (transformed !== href || decodeURIComponent(href).match(matrixLinkify.ELEMENT_URL_PATTERN)) {
                    return null;
                } else {
                    return '_blank';
                }
            } catch (e) {
                // malformed URI
            }
        }
        return null;
    },
};

registerPlugin(Type.RoomAlias, ({ scanner, parser, utils }) => {
    const token = scanner.tokens.POUND as '#';
    return matrixOpaqueIdLinkifyParser({
        scanner,
        parser,
        utils,
        token,
        name: Type.RoomAlias,
    });
});
registerPlugin(Type.GroupId, ({ scanner, parser, utils }) => {
    const token = scanner.tokens.PLUS as '+';
    return matrixOpaqueIdLinkifyParser({
        scanner,
        parser,
        utils,
        token,
        name: Type.GroupId,
    });
});
registerPlugin(Type.UserId, matrixUserIdLinkifyPlugin);
export const linkify = linkifyjs;
