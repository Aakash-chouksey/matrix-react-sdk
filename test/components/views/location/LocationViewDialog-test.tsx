
import React from 'react';
import { mount } from 'enzyme';
import { RoomMember } from 'matrix-js-sdk/src/matrix';
import { LocationAssetType } from 'matrix-js-sdk/src/@types/location';

import LocationViewDialog from '../../../../src/components/views/location/LocationViewDialog';
import { getMockClientWithEventEmitter, makeLocationEvent } from '../../../test-utils';

describe('<LocationViewDialog />', () => {
    const roomId = '!room:server';
    const userId = '@user:server';
    const mockClient = getMockClientWithEventEmitter({
        getClientWellKnown: jest.fn().mockReturnValue({
            "m.tile_server": { map_style_url: 'maps.com' },
        }),
        isGuest: jest.fn().mockReturnValue(false),
    });
    const defaultEvent = makeLocationEvent("geo:51.5076,-0.1276", LocationAssetType.Pin);
    const defaultProps = {
        matrixClient: mockClient,
        mxEvent: defaultEvent,
        onFinished: jest.fn(),
    };
    const getComponent = (props = {}) =>
        mount(<LocationViewDialog {...defaultProps} {...props} />);

    it('renders map correctly', () => {
        const component = getComponent();
        expect(component).toMatchSnapshot();
    });

    it('renders marker correctly for self share', () => {
        const selfShareEvent = makeLocationEvent("geo:51.5076,-0.1276", LocationAssetType.Self);
        const member = new RoomMember(roomId, userId);
        // @ts-ignore cheat assignment to property
        selfShareEvent.sender = member;
        const component = getComponent({ mxEvent: selfShareEvent });
        expect(component.find('SmartMarker').props()['roomMember']).toEqual(member);
    });
});
