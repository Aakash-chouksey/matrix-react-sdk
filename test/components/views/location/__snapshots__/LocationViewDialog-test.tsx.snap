// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LocationViewDialog /> renders map correctly 1`] = `
<LocationViewDialog
  matrixClient={
    MockClientWithEventEmitter {
      "_events": Object {
        "WellKnown.client": [Function],
      },
      "_eventsCount": 1,
      "_maxListeners": undefined,
      "getClientWellKnown": [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": Object {
              "m.tile_server": Object {
                "map_style_url": "maps.com",
              },
            },
          },
        ],
      },
      "isGuest": [MockFunction],
      Symbol(kCapture): false,
    }
  }
  mxEvent={
    Object {
      "content": Object {
        "body": "Found at geo:51.5076,-0.1276 at 2021-12-21T12:22+0000",
        "geo_uri": "geo:51.5076,-0.1276",
        "msgtype": "m.location",
        "org.matrix.msc1767.text": "Found at geo:51.5076,-0.1276 at 2021-12-21T12:22+0000",
        "org.matrix.msc3488.asset": Object {
          "type": "m.pin",
        },
        "org.matrix.msc3488.location": Object {
          "description": "Human-readable label",
          "uri": "geo:51.5076,-0.1276",
        },
        "org.matrix.msc3488.ts": 252523,
      },
      "event_id": "$2",
      "type": "org.matrix.msc3488.location",
    }
  }
  onFinished={[MockFunction]}
>
  <BaseDialog
    className="mx_LocationViewDialog"
    fixedWidth={false}
    hasCancel={true}
    onFinished={[MockFunction]}
  >
    <PosthogScreenTracker />
    <ForwardRef(FocusLockUICombination)
      className="mx_LocationViewDialog"
      lockProps={
        Object {
          "aria-describedby": undefined,
          "aria-labelledby": "mx_BaseDialog_title",
          "onKeyDown": [Function],
          "role": "dialog",
        }
      }
      returnFocus={true}
    >
      <ForwardRef(FocusLockUI)
        as="div"
        autoFocus={true}
        className="mx_LocationViewDialog"
        crossFrame={true}
        disabled={false}
        lockProps={
          Object {
            "aria-describedby": undefined,
            "aria-labelledby": "mx_BaseDialog_title",
            "onKeyDown": [Function],
            "role": "dialog",
          }
        }
        noFocusGuards={false}
        persistentFocus={false}
        returnFocus={true}
        sideCar={[Function]}
      >
        <div
          data-focus-guard={true}
          key="guard-first"
          style={
            Object {
              "height": "0px",
              "left": "1px",
              "overflow": "hidden",
              "padding": 0,
              "position": "fixed",
              "top": "1px",
              "width": "1px",
            }
          }
          tabIndex={0}
        />
        <SideEffect(FocusWatcher)
          autoFocus={true}
          crossFrame={true}
          disabled={false}
          id={Object {}}
          observed={
            <div
              aria-labelledby="mx_BaseDialog_title"
              class="mx_LocationViewDialog"
              data-focus-lock-disabled="false"
              role="dialog"
            >
              <div
                class="mx_Dialog_header mx_Dialog_headerWithCancel"
              >
                <h2
                  class="mx_Heading_h2 mx_Dialog_title"
                  id="mx_BaseDialog_title"
                />
                <div
                  aria-label="Close dialog"
                  class="mx_AccessibleButton mx_Dialog_cancelButton"
                  role="button"
                  tabindex="0"
                />
              </div>
              <div
                class="mx_Map mx_LocationViewDialog_map"
                id="mx_Map_mx_LocationViewDialog_$2"
              >
                <div
                  class="mx_Marker mx_Marker_defaultColor"
                  id="mx_LocationViewDialog_$2-marker"
                >
                  <div
                    class="mx_Marker_border"
                  >
                    <div
                      class="mx_Marker_icon"
                    />
                  </div>
                </div>
                <div
                  class="mx_ZoomButtons"
                >
                  <div
                    class="mx_AccessibleButton mx_ZoomButtons_button"
                    data-test-id="map-zoom-in-button"
                    role="button"
                    tabindex="0"
                    title="Zoom in"
                  >
                    <div
                      class="mx_ZoomButtons_icon"
                    />
                  </div>
                  <div
                    class="mx_AccessibleButton mx_ZoomButtons_button"
                    data-test-id="map-zoom-out-button"
                    role="button"
                    tabindex="0"
                    title="Zoom out"
                  >
                    <div
                      class="mx_ZoomButtons_icon"
                    />
                  </div>
                </div>
              </div>
            </div>
          }
          onActivation={[Function]}
          onDeactivation={[Function]}
          persistentFocus={false}
          returnFocus={[Function]}
          shards={Array []}
          sideCar={
            Object {
              "assignMedium": [Function],
              "assignSyncMedium": [Function],
              "options": Object {
                "async": true,
                "ssr": false,
              },
              "read": [Function],
              "useMedium": [Function],
            }
          }
        >
          <FocusWatcher
            autoFocus={true}
            crossFrame={true}
            disabled={false}
            id={Object {}}
            observed={
              <div
                aria-labelledby="mx_BaseDialog_title"
                class="mx_LocationViewDialog"
                data-focus-lock-disabled="false"
                role="dialog"
              >
                <div
                  class="mx_Dialog_header mx_Dialog_headerWithCancel"
                >
                  <h2
                    class="mx_Heading_h2 mx_Dialog_title"
                    id="mx_BaseDialog_title"
                  />
                  <div
                    aria-label="Close dialog"
                    class="mx_AccessibleButton mx_Dialog_cancelButton"
                    role="button"
                    tabindex="0"
                  />
                </div>
                <div
                  class="mx_Map mx_LocationViewDialog_map"
                  id="mx_Map_mx_LocationViewDialog_$2"
                >
                  <div
                    class="mx_Marker mx_Marker_defaultColor"
                    id="mx_LocationViewDialog_$2-marker"
                  >
                    <div
                      class="mx_Marker_border"
                    >
                      <div
                        class="mx_Marker_icon"
                      />
                    </div>
                  </div>
                  <div
                    class="mx_ZoomButtons"
                  >
                    <div
                      class="mx_AccessibleButton mx_ZoomButtons_button"
                      data-test-id="map-zoom-in-button"
                      role="button"
                      tabindex="0"
                      title="Zoom in"
                    >
                      <div
                        class="mx_ZoomButtons_icon"
                      />
                    </div>
                    <div
                      class="mx_AccessibleButton mx_ZoomButtons_button"
                      data-test-id="map-zoom-out-button"
                      role="button"
                      tabindex="0"
                      title="Zoom out"
                    >
                      <div
                        class="mx_ZoomButtons_icon"
                      />
                    </div>
                  </div>
                </div>
              </div>
            }
            onActivation={[Function]}
            onDeactivation={[Function]}
            persistentFocus={false}
            returnFocus={[Function]}
            shards={Array []}
            sideCar={
              Object {
                "assignMedium": [Function],
                "assignSyncMedium": [Function],
                "options": Object {
                  "async": true,
                  "ssr": false,
                },
                "read": [Function],
                "useMedium": [Function],
              }
            }
          />
        </SideEffect(FocusWatcher)>
        <div
          aria-labelledby="mx_BaseDialog_title"
          className="mx_LocationViewDialog"
          data-focus-lock-disabled={false}
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          role="dialog"
        >
          <div
            className="mx_Dialog_header mx_Dialog_headerWithCancel"
          >
            <Heading
              className="mx_Dialog_title"
              id="mx_BaseDialog_title"
              size="h2"
            >
              <h2
                className="mx_Heading_h2 mx_Dialog_title"
                id="mx_BaseDialog_title"
              />
            </Heading>
            <AccessibleButton
              aria-label="Close dialog"
              className="mx_Dialog_cancelButton"
              element="div"
              onClick={[Function]}
              role="button"
              tabIndex={0}
            >
              <div
                aria-label="Close dialog"
                className="mx_AccessibleButton mx_Dialog_cancelButton"
                onClick={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                role="button"
                tabIndex={0}
              />
            </AccessibleButton>
          </div>
          <Map
            centerGeoUri="geo:51.5076,-0.1276"
            className="mx_LocationViewDialog_map"
            id="mx_LocationViewDialog_$2"
            interactive={true}
            onError={[Function]}
          >
            <div
              className="mx_Map mx_LocationViewDialog_map"
              id="mx_Map_mx_LocationViewDialog_$2"
              onClick={[Function]}
            >
              <SmartMarker
                geoUri="geo:51.5076,-0.1276"
                id="mx_LocationViewDialog_$2-marker"
                map={
                  MockMap {
                    "_events": Object {
                      "error": [Function],
                    },
                    "_eventsCount": 1,
                    "_maxListeners": undefined,
                    "addControl": [MockFunction],
                    "removeControl": [MockFunction],
                    "setCenter": [MockFunction] {
                      "calls": Array [
                        Array [
                          Object {
                            "lat": 51.5076,
                            "lon": -0.1276,
                          },
                        ],
                      ],
                      "results": Array [
                        Object {
                          "type": "return",
                          "value": undefined,
                        },
                      ],
                    },
                    "setStyle": [MockFunction],
                    "zoomIn": [MockFunction],
                    "zoomOut": [MockFunction],
                    Symbol(kCapture): false,
                  }
                }
              >
                <ForwardRef
                  id="mx_LocationViewDialog_$2-marker"
                >
                  <div
                    className="mx_Marker mx_Marker_defaultColor"
                    id="mx_LocationViewDialog_$2-marker"
                  >
                    <div
                      className="mx_Marker_border"
                    >
                      <div
                        className="mx_Marker_icon"
                      />
                    </div>
                  </div>
                </ForwardRef>
              </SmartMarker>
              <ZoomButtons
                map={
                  MockMap {
                    "_events": Object {
                      "error": [Function],
                    },
                    "_eventsCount": 1,
                    "_maxListeners": undefined,
                    "addControl": [MockFunction],
                    "removeControl": [MockFunction],
                    "setCenter": [MockFunction] {
                      "calls": Array [
                        Array [
                          Object {
                            "lat": 51.5076,
                            "lon": -0.1276,
                          },
                        ],
                      ],
                      "results": Array [
                        Object {
                          "type": "return",
                          "value": undefined,
                        },
                      ],
                    },
                    "setStyle": [MockFunction],
                    "zoomIn": [MockFunction],
                    "zoomOut": [MockFunction],
                    Symbol(kCapture): false,
                  }
                }
              >
                <div
                  className="mx_ZoomButtons"
                >
                  <AccessibleButton
                    className="mx_ZoomButtons_button"
                    data-test-id="map-zoom-in-button"
                    element="div"
                    onClick={[Function]}
                    role="button"
                    tabIndex={0}
                    title="Zoom in"
                  >
                    <div
                      className="mx_AccessibleButton mx_ZoomButtons_button"
                      data-test-id="map-zoom-in-button"
                      onClick={[Function]}
                      onKeyDown={[Function]}
                      onKeyUp={[Function]}
                      role="button"
                      tabIndex={0}
                      title="Zoom in"
                    >
                      <div
                        className="mx_ZoomButtons_icon"
                      />
                    </div>
                  </AccessibleButton>
                  <AccessibleButton
                    className="mx_ZoomButtons_button"
                    data-test-id="map-zoom-out-button"
                    element="div"
                    onClick={[Function]}
                    role="button"
                    tabIndex={0}
                    title="Zoom out"
                  >
                    <div
                      className="mx_AccessibleButton mx_ZoomButtons_button"
                      data-test-id="map-zoom-out-button"
                      onClick={[Function]}
                      onKeyDown={[Function]}
                      onKeyUp={[Function]}
                      role="button"
                      tabIndex={0}
                      title="Zoom out"
                    >
                      <div
                        className="mx_ZoomButtons_icon"
                      />
                    </div>
                  </AccessibleButton>
                </div>
              </ZoomButtons>
            </div>
          </Map>
        </div>
        <div
          data-focus-guard={true}
          style={
            Object {
              "height": "0px",
              "left": "1px",
              "overflow": "hidden",
              "padding": 0,
              "position": "fixed",
              "top": "1px",
              "width": "1px",
            }
          }
          tabIndex={0}
        />
      </ForwardRef(FocusLockUI)>
    </ForwardRef(FocusLockUICombination)>
  </BaseDialog>
</LocationViewDialog>
`;
